csrc/python_frontend/python_bindings.cpp

```diff
@@ -462,103 +462,56 @@ void initNvFuserPythonBindings(PyObject* module) {
-                {self.recordingState(out())},
-                serde::RecordType_ConstantInt,
-                val,
-                dtype));
-            return out;
-          },
-          py::arg("val"),
-          py::arg("dtype") = DataType::Int,
-          py::return_value_policy::reference)
       .def(
           "define_scalar",
           [](FusionDefinition& self,
              PrimDataType dtype = DataType::Double) -> Scalar {
-            FUSER_PERF_SCOPE("FusionDefinition.define_scalar");
+            FUSER_PERF_SCOPE("FusionDefinition.define_scalar (input_specific)");
             TORCH_CHECK(
                 !self.completed(),
                 "Attempting to add to a completed definition!");
             Scalar out = self.defineScalar();
-            self.defineRecord(
-                new ScalarRecord({self.recordingState(out())}, dtype));
+            self.defineRecord(new ScalarRecord<std::nullptr_t>(
+                {self.recordingState(out())},
+                serde::RecordType_ScalarInput,
+                nullptr,
+                dtype,
+                true));
             return out;
           },
           py::arg("dtype") = DataType::Double,
           py::return_value_policy::reference);
 
+// This is the canonical version of define_scalar, therefore, it is possible
+// to also define a scalar input with type Null/std::nullptr_t
+// The ScalarRecord will be responsible for checking if is_input is valid.
+#define NVFUSER_PYTHON_BINDING_SCALAR(Nvfuser_DType, CType)                    \
+  fusion_def.def(                                                              \
+      "define_scalar",                                                         \
+      [](FusionDefinition& self, CType val, PrimDataType dtype, bool is_input) \
+          -> Scalar {                                                          \
+        FUSER_PERF_SCOPE("FusionDefinition.define_scalar");                    \
+        Scalar out = self.defineScalar();                                      \
+        self.defineRecord(new ScalarRecord<CType>(                             \
+            {self.recordingState(out())},                                      \
+            serde::mapToSerdeScalarRecordType(Nvfuser_DType),                  \
+            val,                                                               \
+            dtype,                                                             \
+            is_input));                                                        \
+        return out;                                                            \
+      },                                                                       \
+      py::arg("val"),                                                          \
+      py::arg("dtype") = Nvfuser_DType,                                        \
+      py::arg("is_input") = false,                                             \
```

Comment:
How should we interpret this argument? If I understand correctly, we should treat it as an input but hold onto the constant value to insert as an input, so that changing this constant doesn't cause cache misses. That probably warrants some text in the docstring.